rfcs/0001-inline-expression-value-logging.md

```diff
@@ -0,0 +1,191 @@
+> you should avoid having uses of it in version control for longer
+> periods. Use cases involving debug output that should be added to
+> version control may be better served by macros such as `debug!` from
+> the `log` crate.
+
+Indeed, a major point of the `log` package and `Level`s, is the
+ability to keep `debug!` and `trace!` logging in place for further
+use, including by other contributors, without paying a cost for
+unlogged messages in release builds.
+
+It follows that for projects that _do already have_ `log` and an
+output logger dependency and configuration, and particularly with some
+debug/trace logging already in place, use of `std::dbg!` would be
+unwelcome in PRs, and likely less productive than a `log`-based
+equivalent. For parity of convenience, the `log` API needs an
+extension for logging expressions and passing through values inline.
+
+# Detailed Design
+
+In addition to the existing set of _formatted logging_ macros, by
+level, e.g. `trace!`, `debug!`, etc., we add a new set of _inline
+expression and value logging_ macros, with a “_-v_” suffix: `tracev!`,
+`debugv!`, and so forth for all levels. The _-v_ macros take a _single_
+expression argument, which is evaluated exactly once, regardless of if
+the logging level is enabled or not, and returned:
+
+```rust
+let n = 12;
+let m = debugv!(n / 2) - 1;
+//      ^-- debug log message: "n / 2 = 6"
+assert_eq!(m, 5);
+```
+
+The _default_ format string for the _-v_ macros is `"{} = {:?}"`,
+where the `stringify!`-ed expression and resulting value are passed,
+in that order.  Note that the `std::dbg!` macro currently uses `"{} =
+{:#?}"`—the value is "pretty-printed", potentially over multiple
+lines.  Given the line-orientation of logging output, the default
+format for the _-v_ macros avoids this. However, a custom format may
+also be passed before the expression, which adds considerably more
+output flexibility:
+
+```rust
+let i = 32;
+tracev!("{} = {}", i);            // use `Display` instead of `Debug`
+tracev!("{} = {:x}", i);          // hexadecimal format value
+tracev!("{} = {:#?}", i);         // use pretty, multi-line format
+
+let rem = infov!("{1:?} remaining ({0})", deadline - Instance::now());
+                                  // value first, with additional context
+```
+
+For symmetry with the existing `log!` macro, a `logv!` is also
+included, which allows passing a `Level` as parameter.
+
+Finally, like all the other public logging macros, the _-v_ macros
+allows overriding the default module-path target with a string
+literal:
+
+```rust
+let i = 33;
+let j = warnv!(target: "maths", (i-1) / 2);
+assert_eq!(j, 16);
+```
+
+See also the [implementation PR], which includes guide level
+documentation, in the form of rustdoc with doc-tests.
+
+# Alternatives
+
+## Multiple expression support
+
+[RFC 2173] included multiple-expression support for `std::dbg!` but
+was closed, at least in part due to perceived complexity with multiple
```

Comment:
Thanks @Centril. I will refine this alternative subsection to (a) presume less about why RFC 2173 was superseded; (b) point out that _both_ multiple arguments _and_ a  custom format string are at least somewhat at odds, and (c) note that the following still works, as currently designed, by explicitly passing a tuple:

``` rust
let i = 16;
assert_eq!(debugv!((i/4, i/8)), (4, 2));
assert_eq!(debugv!("{} = {:?}", (i/4, i/8)), (4, 2));
```

Regarding your multiple argument example macro: It would be surprising and undesirable in a _log_ context, I think, to write more than one log record for a single _-v_ macro invocation. To "lock stderr only once" sounds like a closely related goal.

Reply:
Thanks @dekellum; I agree with you about the distinction between logging and debugging wrt. the usefulness of multiple arguments; `dbg!(a, b, c)` seems more useful than `debugv!(a, b, c)`. Fwiw, I think this RFC is a nice addition. :)

Reply:
Updated this alternatives subsection, per above, and with further elaboration.